File: DSA/Vaishnavi_590017838/Day_16/Question1.cpp
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        ListNode dummy(0);
        dummy.next = head;
        ListNode* current = &dummy;

        while(current->next != NULL && current->next->next != NULL) {
            ListNode* first = current->next;
            ListNode* second = first->next;
            first->next = second->next;
            second->next = first;
            current->next = second;
            current = first;
        }
        return dummy.next;
    }
};

File: DSA/RiddhiSaxena_590022853/Day_17/question1.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head):
        if head is None or head.next is None:
            return True

        # Step 1: Find the middle of the linked list
        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # Step 2: Reverse the second half
        prev = None
        current = slow

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        # Step 3: Compare both halves
        first = head
        second = prev

        while second:
            if first.val != second.val:
                return False
            first = first.next
            second = second.next

        return True